src/transformers/models/wav2vec2/modeling_wav2vec2.py

```diff
@@ -1511,8 +1514,10 @@ def _mask_hidden_states(
                 attention_mask=attention_mask,
                 min_masks=self.config.mask_time_min_masks,
             )
-            mask_time_indices = torch.tensor(mask_time_indices, device=hidden_states.device, dtype=torch.bool)
-            hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
+            mask_time_indices = mask_time_indices.to(device=hidden_states.device)
+            temp_hidden_states = hidden_states.clone()
```

Comment:
Also not sure why we `clone` here

Reply:
while using this code```hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)``` I am facing this error:
Output 0 of CompiledFunctionBackward is a view and is being modified inplace. This view is the output of a 
function that returns multiple views. Such functions do not allow the output views to be modified inplace. You should replace the inplace operation by an out-of-place one.

Reply:
Okay interesting, but it works when we do:
```python
hidden_size[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
```
on line 1505 above?

I wonder if we can open an issue on the PyTorch repo and see if we can get some clarify here? Not sure exactly what the best protocol is, but would ideally like to avoid cloning the entire tensor if possible

Reply:
The following code works:
```
temp_hidden_states = hidden_states.clone()
temp_hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
hidden_states = temp_hidden_states
```
 The following code throws above mentioned error:
```
hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
```
The following also throws the same error:
```
temp_hidden_states = hidden_states
temp_hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)
hidden_states = temp_hidden_states
```

Can it be because the gradient for hidden_states has the hidden_size in the backward propagation graph and masked_spec_embed also has hidden_size in its backward propagation graph. And this could be an inplace operation?

Reply:
Quite possibly - that sounds like reasonable rationale for this behaviour. cc'ing @fxmarty - have you come across such an error when compiling with Torch 2.0? Torch is complaining that we're doing an index update for a Torch tensor in-place, but not really sure if cloning the tensor is the best option here, or whether we just need to make it non-contiguous?

Reply:
The issue looks to be that `temp_hidden_states[mask_time_indices] = self.masked_spec_embed.to(hidden_states.dtype)` is an inplace operation, and that autograd does not like inplace ops. Maybe [`torch.where`](https://pytorch.org/docs/stable/generated/torch.where.html) would work?

Reply:
```torch.where``` command is working with torch compile. 
The following can be used inplace of ```clone```.
```
expanded_embed = self.masked_spec_embed.unsqueeze(0).unsqueeze(0)
hidden_states = torch.where(mask_time_indices.unsqueeze(-1), expanded_embed, hidden_states)
```

Reply:
I just saw this as @sanchit-gandhi mentioned this to me. I think using `Tensor.masked_fill()` can be slightly faster, while still being an out-of-place op?

Reply:
The ```Tensor.masked_fill()``` seems to only take ```float``` or ```Tensor``` value, I couldnot place ```embeded_mask```  ```1D``` vector in ```masked_fill```.